Fix `meetsAgeRequirements` for min/max dates

francoisl opened the pull request:
### Details
When we check that someone's date of birth is in the `150 years - 18 years` range, we currently do a comparison that:
- uses the time of day
- excludes the min and max dates
- Uses the same error message for both >150 years and <18 years conditions

This change uses splits `meetsAgeRequirements` in two different checks, so that we can use a different error message for each, and also allows the min/max dates to be selected.

### Fixed Issues
$ https://github.com/Expensify/App/issues/17426
$ https://github.com/Expensify/App/issues/17981

### Tests / QA
1. Create a workspace
2. Go to Settings > Workspaces > [Select your workspace] > Issue cards > Connect Bank Account > Connect manually
3. Fill out the 1 and 2 steps of the form
4. In the "Personal information" step, choose the oldest possible date for the date of birth selector (i.e. exactly 150 years ago)
5. Make sure you don't get an error message
6. [On desktop and web only] Use your keyboard to manually change the date to an older date (e.g. press arrow down after selecting the day or month)
7. After unselecting the date input, make sure a validation error appears with the message `Please enter a valid date of birth` (or `Por favor, introduce una fecha de nacimiento válida` in Spanish)
8. Change the date to the most recent possible date - 18 years ago
9. Make sure you don't get an error
10. [On desktop and web only] Use your keyboard to manually change the date to a more recent date (e.g. press arrow up after selecting the day or month)
7. After unselecting the date input, make sure a validation error appears with the message `Must be over 18 years old` (or `Debe ser mayor de 18 años` in Spanish)

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A


### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

- Oldest possible date

<img width="416" alt="image" src="https://user-images.githubusercontent.com/2229301/232176208-a9d1dac2-7bc2-4b9c-a2ea-2bb41a123930.png"><img width="384" alt="image" src="https://user-images.githubusercontent.com/2229301/232176204-7df081cd-1beb-4024-90c8-26b684eabcde.png">

- Changing the date manually to an older date

https://user-images.githubusercontent.com/2229301/234134090-10d7b666-43e8-48be-8a09-a619b54916b5.mov


- Most recent possible date

<img width="398" alt="image" src="https://user-images.githubusercontent.com/2229301/232176236-557c6a14-d56c-49f2-b8bf-e24f707ebd40.png"><img width="393" alt="image" src="https://user-images.githubusercontent.com/2229301/232176241-478d479f-e93a-47dd-b835-49beeda72963.png">


- Changing the date manually to a more recent date


https://user-images.githubusercontent.com/2229301/234134024-ce97d39d-697e-4535-96a7-c2f87ed6f0af.mov


</details>

<details>
<summary>Mobile Web - Chrome</summary>

- Min date
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176696-47c3d0c7-40fe-4e88-a14d-ac79ddbb89ad.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176686-2c449a62-2f10-4f1a-a6f2-52b9848d5b60.png">

- Max date
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176673-65ea4ec2-8aab-4a35-9dda-aca6fe3bc0a9.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176677-5487d2d8-69f5-4695-b27e-3aea20ce54ec.png">


</details>

<details>
<summary>Mobile Web - Safari</summary>

Safari doesn't support restrictions in the date selector https://github.com/Expensify/App/issues/16691#issuecomment-1498863753, but you can still test selecting the min/max dates and make sure there are no error


<img width="395" alt="image" src="https://user-images.githubusercontent.com/2229301/232176512-dec8a14d-e277-4af5-b79d-fbc66c8e904e.png">


Current max date at time of writing this:
<img width="445" alt="image" src="https://user-images.githubusercontent.com/2229301/232176456-fdc02e60-5e82-4b11-8836-76fbdb22525d.png">

Max date + 1 day: error as expected
<img width="461" alt="image" src="https://user-images.githubusercontent.com/2229301/232176477-4ac02cb2-53fe-46d6-8b67-bf93ccf3de2f.png">


</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/2229301/234134200-000d7e34-1ebf-4270-9923-ce0c6ae3580d.mov


</details>

<details>
<summary>iOS</summary>

<img width="414" alt="Screenshot 2023-04-24 at 2 42 31 PM" src="https://user-images.githubusercontent.com/2229301/234134369-c9cbc673-d10b-4b69-a83c-7935c1f51bd8.png">
<img width="420" alt="Screenshot 2023-04-24 at 2 41 50 PM" src="https://user-images.githubusercontent.com/2229301/234134371-e133ea87-f138-4565-b834-1f9f476e75e6.png">


</details>

<details>
<summary>Android</summary>

- Min date

<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176628-21d2b473-8a8c-439c-a78d-d6c6eb20c4b5.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176634-6f09faea-bbcf-43d0-b21f-12322c13d793.png">


- Max date

<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176616-0d117a69-6ff1-4440-8ab6-c00c11678301.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2229301/232176618-640ac069-79a2-4e29-9532-094457c88f79.png">


</details>

francoisl committed:
Fix `meetsAgeRequirements` for min/max dates

francoisl commented:
Will CP to staging because this is blocking testing one PR of the current checklist.

Gonals commented:
Rerunning pullerBear, as I'm leaving on parental leave!

Gonals commented:
Rerunning pullerBear, as I'm leaving on parental leave!

s77rt commented:
Can you please fix this for `getAgeRequirementError` too
https://github.com/Expensify/App/blob/af3ae131b866365661dc52c995e17b572037fd63/src/libs/ValidationUtils.js#L226
The last parameter should be a string, although the empty array also seems to work.

francoisl commented on src/libs/ValidationUtils.js:
```diff
@@ -203,7 +203,9 @@ function meetsAgeRequirements(date) {
-    return testDate.isValid() && testDate.isBetween(oneHundredFiftyYearsAgo, eighteenYearsAgo);
+
+    // Only compare the dates (ignore the time), and make the comparison inclusive of the start and end dates
+    return testDate.isValid() && testDate.isBetween(oneHundredFiftyYearsAgo, eighteenYearsAgo, 'day', '[]');
```
You mean it doesn't work in that the message is technically wrong? Yeah fair point, I'm trying to thing what's the best way to update it:

- Either we can update just the message to be something like "Must be over 18 years old and less than 150 years old", or
- Split the function `meetsAgeRequirements` into two functions, e.g. `isBelowAgeRequirement` and `isOverAgeRequirement` but that sounds overkill, or
- Change the function `meetsAgeRequirements` to return an int, for example `-1` if the age is < 18, `0` if it's fine, `1` if it's over 150 years

Any preference or alternative ideas?

francoisl replied:
> correct because this is the main issue we're trying to fix and makes user confused.

Oooh, admittedly I read the issue too fast and thought the problem that QA reported was that we were showing the error if you selected the oldest possible date in the date selector. But yes you're right, and we should use use a different error message then.

Splitting `meetsAgeRequirements` into two functions works for me. I'm checking about the new error message now because it sounds weird to say you have to be less than 150 years old.

francoisl committed:
Merge branch 'main' of github.com:Expensify/App into francois-dateRequirementInclusiveComparison

francoisl committed:
Split age requirement function into two functions

Introduce `meetsMinimumAgeRequirement()` and `meetsMaximumAgeRequirement()` to replace `meetsAgeRequirements()`.
The two functions allow us to show a different error message if a given date is too recent or too far back. For
the case a given date is too old for an age, we're reusing the existing invalid date error message for simplicity.

francoisl commented:
Updated, I split `meetsAgeRequirements` into two functions. For the case where the date is too old, we're reusing the existing generic error message "Please enter a valid date of birth" as suggested [here](https://github.com/Expensify/App/issues/17426#issuecomment-1514515658).

0xmiroslav reviewed (commented):
Should we also handle these pages here or out of scope?

<img width="367" alt="Screenshot 2023-04-25 at 9 33 32 PM" src="https://user-images.githubusercontent.com/97473779/234383731-1b4eff27-3666-438b-8bea-6c9ea7c7434c.png">
<img width="364" alt="Screenshot 2023-04-25 at 9 33 46 PM" src="https://user-images.githubusercontent.com/97473779/234383740-f9d68a74-c6bc-4af3-ac90-dc850e4c6a75.png">

francoisl committed:
Fix duplicate word in function doc

francoisl commented:
> Should we also handle these pages here or out of scope?

These are out of scope, AFAIK we don't have a date restriction on the incorporation date.

0xmiroslav reviewed (approved):
LGTM 🎉

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4815996583) to staging by https://github.com/marcochavezf in version: 1.3.7-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4828174056) to production by https://github.com/jasperhuangg in version: 1.3.7-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
